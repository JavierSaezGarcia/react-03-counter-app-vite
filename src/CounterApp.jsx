import { useState } from 'react';
import PropTypes from 'prop-types';

export const CounterApp = ({ value }) => {
    // En useState ponemos el valor inicial
    // El primer valor desestructurado del array es la variable que vamos a operar y 
    // que la llamaremos como queramos (contador) y el segundo parametro es una funcion que se ejecuta cuando cambia el valor
    // y que llamaremos set y el nombre en camelcase de la  variable  
    const [ counter, setCounter ] = useState(value);
    const handleAdd = () => { setCounter( counter + 1) } 
       // setContador( (c) => c + 1 )  // Otra forma de hacerlo según necesitemos    
    const handleSubstract = () => { setCounter( counter - 1) }
        // setContador( (c) => c - 1 )  // Otra forma de hacerlo según necesitemos    
    const handleReset = () => { setCounter( value ) }
    return (
        <>
            <h1>CounterApp</h1>
            <h2> { counter } </h2>
            <button onClick={ handleAdd }> +1 </button>
            <button onClick={ handleSubstract }> -1 </button>
            <button aria-label="btn-reset" onClick={ handleReset }> Reset </button>
        </>
    )
}

CounterApp.propTypes = {
    value: PropTypes.number.isRequired
}
