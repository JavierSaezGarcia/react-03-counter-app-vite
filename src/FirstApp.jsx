import PropTypes from 'prop-types';

// const newMessage = 'Javier'; // todo lo que este fuera de este componente no se va a renderizar en el html
// const masMessage = [1,2,3,4,5,6,7,8,9];

// const otroMessage = {
//   title: 'Hola mundo desde objeto',
//   message: 'soy un mensaje'
// }
// Si esta funcion suma fuera asincrona (const suma = async () => {}) (lo que hace es que no se ejecutaria el codigo hasta que llegue la respuesta)tendriamos un problema de que no se ejecutaria el codigo hasta que llegue la respuesta
// Por este motivo no es conveniente hacerlo
// const suma = (a,b) => { 
//   return a + b;
// }

// NO PODEMOS MANDAR UN OBJETO PERO SI LO NECESITAMOS PODEMOS MANDAR UN OBJETO SERIALIZADO CON json.stringify(OBJETO)



export const FirstApp = ( { title, subTitle, name } ) => {
  // evitar poner funciones dentro de las llaves de un componente si es posible ya que cada vez que cambie asignara un espacio en la memoria
  // y si esta fuera no, porque solo se redibuja lo que esta dentro de la funcion
  // const suma = (a,b) => { 
  //   return a + b;
  // }

  if( !title){
    throw new Error('El titulo es obligatorio');
  }
  
  
  return ( // forma abreviada de <Fragments></Fragments> que son como divs <></> pero sin hacer un elemento html
    <>
      <h1 data-testid="test-title">{ title }</h1>
      <h2>{ subTitle }</h2>
      <h2>{ subTitle }</h2>
      <h2>{ subTitle }</h2>
      <h2>{ subTitle }</h2>
      <h3>{ name }</h3>
      {/* <h3>{ masMessage }</h3> 
      <h4>{ otroMessage.title  } </h4> */}
      {/* <h4>{ otroMessage.message }</h4> */}
      {/* <code>Pero de esta forma es mejor: { JSON.stringify(otroMessage) }</code>
      <br />
      <h3>La suma  es: { suma(5,2) }</h3>
      <br />
      <cite>Mis apellidos son:  { apellidos }</cite> */}
      
    </>
    
  )
}
// Los default props entran antes que el resto de props
// FirstApp.defaultProps = {
//   title: 'No hay titulo',
//   subTitle: 'No hay subtitulo',
//   name: 'No hay nombre'
// }

FirstApp.propTypes = {
  title: PropTypes.string.isRequired,
  subTitle: PropTypes.string, 
  name: PropTypes.string
}

