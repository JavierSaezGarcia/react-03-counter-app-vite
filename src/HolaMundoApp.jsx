
// function App() {
//     return (<h1>Hola, querido mundo</h1>);
// }

// export default App;

// export function App() {
//     return (<h1>Hola, querido mundo</h1>);
// }


// export const HolaMundoApp = () => {
//   return (
//     <h1>Hola Mundo App!!</h1>
//   )
// }

