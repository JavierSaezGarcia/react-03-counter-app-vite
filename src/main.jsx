import React from 'react';
import ReactDOM from 'react-dom/client';
// import HolaCaracola from './HolaMundoApp';
// import {  HolaMundoApp } from './HolaMundoApp';
// import { FirstApp } from './FirstApp';
import './styles.css';
import { CounterApp } from './CounterApp';

ReactDOM.createRoot(document.getElementById('root')).render(
    <React.StrictMode>          
        {/* <FirstApp title="Hola, soy Goku" subTitle="No me gusta Vegetta" name="Javier Saez" /> */}
        <CounterApp value={100} />    

    </React.StrictMode>
);




