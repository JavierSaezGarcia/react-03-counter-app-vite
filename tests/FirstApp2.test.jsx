/* eslint-disable no-undef */
import { render, screen } from "@testing-library/react";
import { FirstApp } from "../src/FirstApp";


describe('Testing <FirstApp />', () => {
    // Constants
    const title = 'Hola, soy Goku';
    const subTitle = 'No me gusta Vegetta';
    // const name = 'Javier Saez'

    test('should match with the snapshot', () => {
        const { container } = render(<FirstApp title={title} />);
        expect(container).toMatchSnapshot();
    });
    test('should display the message "Hola, soy Goku"', () => {        
        render(<FirstApp title={title} />);        
        expect(screen.getByText(title)).toBeTruthy();
        // screen.debug();
    });
    test('should display the title in <h1> tag', () => {
        render(<FirstApp title={title} />);
        expect(screen.getByRole('heading', { level: 1 }).innerHTML).toContain(title);
    });
    test('should display the subtitle', () => {
        render(<FirstApp title={title} subTitle={subTitle} />);
        // expect(screen.getByRole('heading', { level: 2 }).innerHTML).toContain(subTitle);
        expect( screen.getAllByText(subTitle).length ).toBe(4);
    });
});

