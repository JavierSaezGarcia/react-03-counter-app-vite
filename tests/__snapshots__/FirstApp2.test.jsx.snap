// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing <FirstApp /> should match with the snapshot 1`] = `
<div>
  <h1
    data-testid="test-title"
  >
    Hola, soy Goku
  </h1>
  <h2 />
  <h2 />
  <h2 />
  <h2 />
  <h3 />
</div>
`;
